nservicebus/transactional-session/index.md

```diff
@@ -0,0 +1,155 @@
+
+```mermaid
+sequenceDiagram
+    actor User
+    autonumber
+    Note over User: Phase 1
+    activate User
+    User->>TransactionalSession: Open()
+    activate TransactionalSession
+    TransactionalSession->>PendingTransportOperations: new()
+    TransactionalSession->>Storage: BeginTransaction()
+    activate Storage
+    Storage-->>TransactionalSession: transaction
+    deactivate Storage
+    deactivate TransactionalSession
+
+    User->>TransactionalSession: Publish<UserCreated>()
+    activate TransactionalSession
+    TransactionalSession->>PendingTransportOperations: Add()
+    deactivate TransactionalSession
+    activate TransactionalSession
+    deactivate TransactionalSession
+    User->>Storage: Store(user)
+
+    User->>TransactionalSession: Commit()
+    activate TransactionalSession
+    TransactionalSession->>Transport: Dispatch(controlMessage)
+    TransactionalSession->>PendingTransportOperations: ConvertToOutboxOperations
+    activate PendingTransportOperations
+    PendingTransportOperations-->>TransactionalSession: outboxRecord
+    deactivate PendingTransportOperations
+    TransactionalSession->>Storage: Store(outboxRecord)
+    TransactionalSession->>Storage: Complete()
+    deactivate TransactionalSession
+    deactivate User
+
+    Note over ReceivePipeline: Phase 2
+    ReceivePipeline->>Transport: Read()
+    activate ReceivePipeline
+    ReceivePipeline->>Storage: GetOutboxRecord()
+    activate Storage
+    Storage-->>ReceivePipeline: outboxRecord
+    deactivate Storage
+    ReceivePipeline->>Transport: Dispatch(outboxRecord.Operations)
+    ReceivePipeline->>Storage: SetAsDispatched()
+    ReceivePipeline-->>Transport: Ack()
+    deactivate ReceivePipeline
+```
+
+There is no single transaction that spans all the operations. The operations occur in two separate phases:
+
+### Phase 1
+
+1. The user opens a transactional session.
+2. A set of `PendingOperations` is initialized and collects the message operations.
+3. A transaction is started on the storage seam.
+4. The user can execute any required message operations using the transactional session.
+5. The user can store any data using the persistence-specific session, which is accessible through the transactional session.
+6. When all operations are registered, the user calls ´Commit´ on the transactional session.
+7. A control message is dispatched to the local queue.
+8. The message operations are converted and stored into an outbox record.
+9. The transaction is committed, and the outbox record and business data modifications are stored atomically.
+
+### Phase 2
+
+The endpoint receives the control message and processes it as follows:
+   * Find the outbox record.
+   * If it exists, it hasn't been marked as dispatched, and there are pending operations, dispatch those operations, and set the outbox record as dispatched.
+   * If it doesn't exist, processing of the control message is delayed to a later point.
+
+## Important design consideration
```

Comment:
Should we add something about we assume the message system to be highly available?